File: backend/app/routers/geocode.py
import httpx
from fastapi import APIRouter, HTTPException, Query

from app.schemas import ReverseGeocodeResponse

router = APIRouter(prefix="/geocode", tags=["geocode"])

NOMINATIM_BASE = "https://nominatim.openstreetmap.org"
HEADERS = {"Accept-Language": "en", "User-Agent": "CivicPulse-FARM/1.0"}


@router.get("/reverse", response_model=ReverseGeocodeResponse)
async def reverse_geocode(
    lat: float = Query(...),
    lng: float = Query(...),
) -> ReverseGeocodeResponse:
    url = f"{NOMINATIM_BASE}/reverse?format=json&lat={lat}&lon={lng}&addressdetails=1&zoom=18"

    async with httpx.AsyncClient(timeout=15.0) as client:
        response = await client.get(url, headers=HEADERS)

    if response.status_code != 200:
        raise HTTPException(status_code=502, detail="Geocoding service unavailable")

    data = response.json()
    addr = data.get("address") or {}
    parts = [
        addr.get("house_number"),
        addr.get("road") or addr.get("pedestrian") or addr.get("footway"),
        addr.get("suburb") or addr.get("neighbourhood"),
        addr.get("city") or addr.get("town") or addr.get("village"),
        addr.get("state"),
    ]
    parts = [part for part in parts if part]

    if parts:
        address = ", ".join(parts)
    else:
        address = data.get("display_name") or f"{lat:.5f}, {lng:.5f}"

    return ReverseGeocodeResponse(address=address)

File: backend/src/routes/geocode.js
import { Router } from 'express'

const router = Router()
const NOMINATIM_BASE = 'https://nominatim.openstreetmap.org'
const HEADERS = {
  'Accept-Language': 'en',
  'User-Agent': 'PublicEye-Express/1.0',
}

/**
 * GET /api/geocode/reverse - Reverse geocode latitude and longitude to human address
 */
router.get('/reverse', async (req, res, next) => {
  const { lat, lng } = req.query

  if (!lat || !lng) {
    return res.status(400).json({ detail: 'lat and lng query parameters are required' })
  }

  const url = `${NOMINATIM_BASE}/reverse?format=json&lat=${encodeURIComponent(lat)}&lon=${encodeURIComponent(lng)}&addressdetails=1&zoom=18`

  try {
    const response = await fetch(url, { headers: HEADERS, signal: AbortSignal.timeout(15000) })

    if (!response.ok) {
      return res.status(502).json({ detail: 'Geocoding service unavailable' })
    }

    const data = await response.json()
    const addr = data.address || {}

    const parts = [
      addr.house_number,
      addr.road || addr.pedestrian || addr.footway,
      addr.suburb || addr.neighbourhood,
      addr.city || addr.town || addr.village,
      addr.state,
    ].filter(Boolean)

    let address = ''
    if (parts.length > 0) {
      address = parts.join(', ')
    } else {
      const parsedLat = parseFloat(lat)
      const parsedLng = parseFloat(lng)
      address = data.display_name || `${parsedLat.toFixed(5)}, ${parsedLng.toFixed(5)}`
    }

    return res.json({ address })
  } catch (err) {
    return res.status(502).json({ detail: 'Geocoding service unavailable' })
  }
})

export default router
